src/src/extra-providers.ts

```diff
@@ -1,36 +1,64 @@
-import { Injectable, NgZone, StaticProvider, Inject } from '@angular/core';
+import { Injectable, StaticProvider, Inject } from '@angular/core';
 import {
   ɵBrowserPlatformLocation,
   PlatformLocation,
   LocationChangeEvent,
   DOCUMENT,
 } from '@angular/common';
 
+type OnPopStateListener = (event: LocationChangeEvent) => void;
+
+declare const Zone: any;
+
 @Injectable()
 export class SingleSpaPlatformLocation extends ɵBrowserPlatformLocation {
-  private ngZone!: NgZone;
-
   // This is a simple marker that helps us to ignore PopStateEvents
   // that was not dispatched by the browser.
   private skipNextPopState = false;
 
-  private onPopStateListeners: ((event: LocationChangeEvent) => void)[] = [];
+  // The key here is an actual forked `Zone` of some specific application.
+  // We will be able to find the specific zone when application gets destroyed
+  // by application `name`.
+  // The reason of that the `onPopState` method is invoked during `bootstrapModule`
+  // and we can't know what application has invoked it. Why should we know the application
+  // that has invoked `onPopState`? When application gets destroyed in a `sharing dependencies mode`
+  // (when there is a single platform per all applications) we want to remove application
+  // specific `popstate` listeners. E.g. if there are 2 applications:
+  // * shop application adds `popstate` listener
+  // * navbar application adds `popstate` listener
+  // When shop application gets destroyed we want to remove only its `popstate` listener.
+  private zoneToOnPopStateListenersMap = new Map<any, OnPopStateListener[]>();
```

Comment:
Why can't this map have the zoneIdentifier as key? Surely we can extract the zone identifier from the zone and use that as the key? I've read through your comments here and am still unsure about it.

Reply:
Yep, it's not accessible yet because this `onPopState` method is called during `await bootstrapFunction()` and before we set our custom property.